tests/test_campaigns.py

```diff
@@ -0,0 +1,75 @@
+
+
+class TestCampaign(unittest.TestCase):
+    @unittest.skipIf("TESTMODULE" not in os.environ, "TESTMODULE not set")
+    @unittest.skipIf(
+        os.environ["TESTMODULE"].strip("/").endswith("run2_2016_nano_v9"),
+        "Tests for module 'run2_2016_nano_v9' are not implemented yet.",
+    )
+    @unittest.skipIf(
+        os.environ["TESTMODULE"].strip("/").endswith("__pycache__"),
+        "'__pycache__' is not a valid module.",
+    )
+    @classmethod
+    def setUpClass(cls):
+        modpath = os.path.realpath(os.path.dirname(os.path.dirname(__file__)))
+        if modpath not in sys.path:
+            sys.path.append(modpath)
+        mod = os.environ["TESTMODULE"]
+        to_import = os.path.relpath(mod, modpath).replace("/", ".")
+        # Load the module to test
+        cls.module = imp.import_module(to_import)
+
+        # get all submodules
+        all_objects = getattr(cls.module, "__all__", None)
+        cls.submodules = list()
+        if not isinstance(all_objects, list):
+            basename = cls.module.__name__.split(".")[-1]
+            object_name = f"campaign_{basename}"
+            cls.submodules = {object_name: getattr(cls.module, object_name)}
+        else:
+            cls.submodules = {x: getattr(cls.module, x) for x in all_objects}
+
+    def test_campaign_properties(self):
+        # loop through the different campaign objects in the current module
+        for name, submodule in self.submodules.items():
+            # run the test for each submodule
+            with self.subTest(f"testing object {name}"):
+                # make sure the campaign defines a basic set of properties
+                self.assertTrue(hasattr(submodule, "name"))
+                self.assertTrue(hasattr(submodule, "id"))
+                self.assertTrue(hasattr(submodule, "ecm"))
+                self.assertTrue(hasattr(submodule, "bx"))
+
+    def test_campaign_datasets(self):
+        # loop through the different campaign objects in the current module
+        for campaign_name, submodule in self.submodules.items():
+            # run the test for each submodule
+            with self.subTest(f"testing dataset of object '{campaign_name}'"):
+                # make sure the campaign defines datasets in the first place
+                self.assertTrue(hasattr(submodule, "datasets"))
+                datasets = submodule.datasets
+                # loop through the datasets and test their properties
+                for dataset_name, _, dataset in datasets.items():
+                    with self.subTest(f"testing dataset {campaign_name}/{dataset_name}"):
+                        # might want to implement naming conventions here,
+                        # e.g. the name of the dataset is the process name + suffix
+                        self.assertTrue(hasattr(dataset, "name"))
+                        self.assertTrue(hasattr(dataset, "id"))
+                        self.assertTrue(hasattr(dataset, "processes"))
```

Comment:
Yes, this does look more like a test of the `order` object, but I think it's still useful to have in case of unexpected upstream changes. I was also thinking if we should check the list of processes explicitly, but in the end it's probably best to allow for the testing use case you mentioned.

Reply:
I agree that this in rather a check for order functionalities, but I do think this is reasonable since e.g. columnflow requires specific information to run properly, so I think making sure that this information is accessible is important. Thanks @mafrahm for spotting the feature of empty processes - this is so simple and general that I went ahead and included an additional test :+1: